### PR TITLE
feature/auth : NbOAuth2AuthStrategy implementing basic authentication scheme against token endpoints

### PR DESCRIPTION
**Pull request by alain-charles**
#### What it resolves
NbOAuth2Strategy now implements client authentication as specified in [RFC 6749 section 2-3](https://tools.ietf.org/html/rfc6749#section-2.3)

There is a new optional parameter of `NbOAuth2StrategyOption`.
The parameter is `clientAuthMethod`, and is a member of `NbOAuth2ClientAuthMethod ` enum:
- `NONE` (default) : no credentials are sent => No breaking change,
- `BASIC` : credentials are sent in the authorization header
- `REQUEST_BODY`: credentials are sent in the request body

AuthMethod is used (credentials are sent) when accessing to the authServer for :
- Getting token with `authorization_code` grant_type
- Getting token with `password` grant-type
- Getting token with `refresh_token` grant-type

RFC6749 says the client must not authenticate when hitting authorize endpoints, even if asking for a token. So nothing changed here, only clientId is sent in the url.
 
 #### issue resolved
issue #581

### CODE REVIEW

File: src/framework/auth/strategies/oauth2/oauth2-strategy.ts [hunk 1]
```diff
@@ -282,6 +282,20 @@ export class NbOAuth2AuthStrategy extends NbAuthStrategy {
     return this.cleanParams(params);
   }
 
+  protected buildRequestsHttpOptions(): any {
```
Comment (nnixaa) [inline src/framework/auth/strategies/oauth2/oauth2-strategy.ts]: let's make it optional?

File: src/framework/auth/strategies/oauth2/oauth2-strategy.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@
  * Licensed under the MIT License. See License.txt in the project root for license information.
  */
 import { Inject, Injectable } from '@angular/core';
-import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import {HttpClient, HttpErrorResponse, HttpHeaders} from '@angular/common/http';
```
Comment (nnixaa) [inline src/framework/auth/strategies/oauth2/oauth2-strategy.ts]: spaces

File: src/framework/auth/strategies/oauth2/oauth2-strategy.ts (hunk 1 above)
Comment (alain-charles) [reply src/framework/auth/strategies/oauth2/oauth2-strategy.ts]: This makes sense since clientId and clientSecret are optional

Comment (nnixaa) [reply src/framework/auth/strategies/oauth2/oauth2-strategy.ts]: Sorry, my comment was a bit misleading. What I meant is that probably we should add an option to the configuration as you proposed in the PR description indicating if we want to enable the base auth, pass clientid/cliensecret as body parameters or none of the above?

Comment (nnixaa) [reply src/framework/auth/strategies/oauth2/oauth2-strategy.ts]: @alain-charles just checking if you saw my comment :)

**Comment by alain-charles** [comment]
I have seen it in my mailbox

Going to setup the options none / auth / body 

Alain

File: src/framework/auth/strategies/oauth2/oauth2-strategy.spec.ts [hunk 3]
```diff
@@ -9,9 +9,8 @@ import { HttpClientTestingModule, HttpTestingController } from '@angular/common/
 import { RouterTestingModule } from '@angular/router/testing';
 import { ActivatedRoute } from '@angular/router';
 import { NB_WINDOW } from '@nebular/theme';
-
 import { NbOAuth2AuthStrategy } from './oauth2-strategy';
-import { NbOAuth2GrantType, NbOAuth2ResponseType } from './oauth2-strategy.options';
+import {NbOAuth2ClientAuthMethod, NbOAuth2GrantType, NbOAuth2ResponseType} from './oauth2-strategy.options';
```
Comment (nnixaa) [inline src/framework/auth/strategies/oauth2/oauth2-strategy.spec.ts]: spaces :)

**Pull request merged**